File: PO/Project/C++/src/logmanager.cpp
#include "logmanager.h"
#include "animals.h"
#include "plants.h"
#include "organizm.h"
#include "plants.h"
#include <cstdarg>
#include <ncurses.h>
#include <string>

logmanager::logmanager(const char* filepath)
:filepath(filepath), logFile(filepath)
{
    if (logFile.is_open()) 
    {
        logFile << "---!log started!---\n";
    } else {
        // Print an error message if the file couldn't be opened
        std::cerr << "Error: Unable to open log file for writing.\n";
    }

    int max_y, max_x;
    getmaxyx(stdscr, max_y, max_x);

    logWindow = newwin(max_y, max_x * 0.4, 0, max_x * 0.6);

    init_pair(NORMAL, COLOR_WHITE, -1);
    init_pair(SPAWN, COLOR_GREEN, -1);
    init_pair(FIGHT, COLOR_YELLOW, -1);
    init_pair(DEATH, COLOR_RED, -1);
    init_pair(WARNING, COLOR_CYAN, -1);
    init_pair(ABILITY, COLOR_BLUE, -1);
    init_pair(TECHNICAL, COLOR_MAGENTA, -1);

    //box(logWindow, 0, 0);
    scrollok(logWindow, TRUE);
    wrefresh(logWindow);

    LogF(TECHNICAL, "---!log started!---\n\n");
}

std::string logmanager::GetEntityName(organizm* entity)
{
    if (dynamic_cast<player*>(entity) != nullptr)
    {
        return "player";
    }
    else if (dynamic_cast<wolf*>(entity) != nullptr)
    {
        return "wolf";
    }
    else if (dynamic_cast<sheep*>(entity) != nullptr)
    {
        return "sheep";
    }
    else if (dynamic_cast<fox*>(entity) != nullptr)
    {
        return "fox";
    }
    else if (dynamic_cast<turtle*>(entity) != nullptr)
    {
        return "turtle";
    }
    else if (dynamic_cast<antelope*>(entity) != nullptr)
    {
        return "antelope";
    }
    else if (dynamic_cast<grass*>(entity) != nullptr)
    {
        return "grass";
    }
    else if (dynamic_cast<mlecz*>(entity) != nullptr)
    {
        return "mlecz";
    }
    else if (dynamic_cast<guarana*>(entity) != nullptr)
    {
        return "guarana";
    }
    else if (dynamic_cast<wolfberry*>(entity) != nullptr)
    {
        return "wolfberry";
    }
    else if (dynamic_cast<hogweed*>(entity) != nullptr)
    {
        return "hogweed";
    }
    else
    {
        return "ERROR";
    }
}

void logmanager::Log(std::string prompt)
{
    if (!logFile.is_open())
    {
        //error!
        return;
    }

    logFile << prompt << '\n';
    LogF(WARNING, (prompt + "\n\n").c_str());

    return;
}

void logmanager::LogF(int color, const char* prompt, ...)
{
    wattron(logWindow, COLOR_PAIR(color));
    va_list args;
    va_start(args, prompt);
    char buffer[256];
    vsnprintf(buffer, sizeof(buffer), prompt, args);

    std::string logMessage = std::string(buffer);
    wprintw(logWindow, logMessage.c_str());
    wrefresh(logWindow);

    va_end(args);
    wattroff(logWindow, COLOR_PAIR(color));
}


void logmanager::LogTime()
{
    if (!logFile.is_open())
    {
        //error!
        return;
    }

    logFile << "current time is " << time(NULL) << "\n\n";
}

void logmanager::LogCreation(organizm* created)
{
    if (!logFile.is_open())
    {
        //error!
        return;
    }

    logFile << "Created " << GetEntityName(created) << " at y = " << created->GetPosition().y << " x = " << created->GetPosition().x << '\n';

    LogF(SPAWN, "Created %s at y = %d and x = %d\n\n",GetEntityName(created).c_str() , created->GetPosition().y, created->GetPosition().x);
    return;
}

void logmanager::LogCollision(organizm* attaker, organizm* prey)
{
    if (!logFile.is_open())
    {
        //error!
        return;
    }
    
    std::string attaker_name = GetEntityName(attaker).c_str();
    std::string prey_name = GetEntityName(prey).c_str();

    logFile << "Fight log: " << attaker_name << " vs " << prey_name << "\n";

    logFile << "location: y = " << prey->GetPosition().y << " x = " << prey->GetPosition().x << "\n";

    logFile << "round: " << attaker->GetWorld()->GetRound() << '\n';

    logFile << attaker_name << " strength = " << attaker->GetStrength() << ", " << prey_name << " strength = " << prey->GetStrength() << '\n';

    LogF(FIGHT, ("Fight: " + attaker_name + " vs " + prey_name + "\n").c_str());

    return;
}

void logmanager::LogCollisionResult(organizm* winner)
{
    if (!logFile.is_open())
    {
        //error!
        return;
    }

    logFile << "the winner is " << GetEntityName(winner).c_str() << "\n\n";

    //LogF(FIGHT, ("Winner: " + GetEntityName(winner)).c_str());

    return;
}

void logmanager::LogReproductionCollision(organizm* roditiel1, organizm* roditiel2)
{
    if (!logFile.is_open())
    {
        //error!
        return;
    }

    logFile << GetEntityName(roditiel1).c_str() << " and " << GetEntityName(roditiel2).c_str() << " now have a child!\n\n";

    return;
}

void logmanager::LogOrder(std::vector<organizm*> members)
{
    if (!logFile.is_open())
    {
        //error!
        return;
    }

    logFile << "The New Order:\n";

    for (int i = 0; i < members.size(); i++) 
    {
        logFile << i << ") " << GetEntityName(members[i]) << '\n';
    }

    logFile << '\n';

    return;
}

void logmanager::LogStrengthIncrease(organizm* lucky_boy)
{
    if (!logFile.is_open())
    {
        //error!
        return;
    }

    logFile << "Strength of the " << GetEntityName(lucky_boy) << " has been increased by 3, new strength is " << lucky_boy->GetStrength() << '\n';

    LogF(ABILITY, "Strength of the %s has been increased by 3, new strength is %d\n\n", GetEntityName(lucky_boy).c_str(), lucky_boy->GetStrength());

    return;
}

void logmanager::LogOverpopulation(organizm* unlucky_boy, int neighbours)
{
    if (!logFile.is_open())
    {
        //error!
        return;
    }

    logFile << GetEntityName(unlucky_boy).c_str() << " has died because of overpopulation, it had " << neighbours << " neighbours\n\n";

    return;
}

void logmanager::NecroLog(organizm* RIP)
{
    if (!logFile.is_open())
    {
        //error!
        return;
    }

    logFile << GetEntityName(RIP).c_str() << " is dead :(\nIt's age was " << RIP->GetWorld()->GetRound() - RIP->GetBirth() << "\n\n";

    LogF(DEATH, "%s died at (y = %d, x = %d), it's age was %d\n\n", GetEntityName(RIP).c_str(), RIP->GetPosition().y, RIP->GetPosition().x, RIP->GetWorld()->GetRound() - RIP->GetBirth());

    return;
}

logmanager::~logmanager()
{
    if (logFile.is_open()) 
    {
        logFile << "---!log ended!---";
        logFile.close();
    }

    delwin(logWindow);   
}

File: PO/Project/JavaProject/src/Abstractions/LogManager.java
package Abstractions;

import Animals.*;
import Plants.*;

import java.awt.*;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.List;
import javax.swing.*;

public class LogManager extends JTextArea {
    public LogManager(int width, int height) {
        super(height, width);
        setEditable(false);
        append("---!log started!---\n\n");
        setFocusable(false);
    }

    public String getEntityName(Organism entity) {
        if (entity instanceof Player) return "player";
        else if (entity instanceof Wolf) return "wolf";
        else if (entity instanceof Sheep) return "sheep";
        else if (entity instanceof Fox) return "fox";
        else if (entity instanceof Turtle) return "turtle";
        else if (entity instanceof Antelope) return "antelope";
        else if (entity instanceof Grass) return "grass";
        else if (entity instanceof Mlecz) return "mlecz";
        else if (entity instanceof Guarana) return "guarana";
        else if (entity instanceof Wolfberry) return "wolfberry";
        else if (entity instanceof Hogweed) return "hogweed";
        else return "ERROR";
    }

    public void log(LogMessageType type, String prompt, Object... args) {
        // Get the color associated with the LogMessageType
        Color color = type.getColor();

        // Create a formatter to format the message with arguments
        Formatter formatter = new Formatter();
        formatter.format(prompt, args);

        // Set the color of the text
        setForeground(color);

        // Append the formatted message to the text area
        append(formatter.toString());

        // Reset the color to default (black)
        setForeground(Color.BLACK);
    }


    public void logTime() {
        String time = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date());
        log(LogMessageType.WARNING, "Current time is " + time + "\n\n");
    }

    public void logCreation(Organism created) {
        String message = String.format("Created %s at y = %d and x = %d",
                getEntityName(created), created.getPosition().y, created.getPosition().x);
        log(LogMessageType.SPAWN, message + "\n\n");
    }

    public void logCollision(Organism attacker, Organism prey) {
        String attackerName = getEntityName(attacker);
        String preyName = getEntityName(prey);

        String message = String.format("Fight log: %s vs %s\nLocation: y = %d, x = %d\nRound: %d\n%s strength = %d, %s strength = %d",
                attackerName, preyName, prey.getPosition().y, prey.getPosition().x,
                attacker.getWorld().getRound(), attackerName, attacker.getStrength(), preyName, prey.getStrength());
        log(LogMessageType.FIGHT, "Fight: %s vs %s (%d | %d)\n",
                attackerName, preyName, attacker.getStrength(), prey.getStrength());
    }

    public void logCollisionResult(Organism winner) {
        String message = String.format("The winner is %s\n\n", getEntityName(winner));
        log(LogMessageType.FIGHT,message);
    }

    public void logReproductionCollision(Organism parent1, Organism parent2) {
        String message = String.format("%s and %s now have a child!\n\n",
                getEntityName(parent1), getEntityName(parent2));
        log(LogMessageType.SPAWN,message);
    }

    public void logOrder(List<Organism> members) {
        log(LogMessageType.WARNING,"The New Order:");
        for (int i = 0; i < members.size(); i++) {
            log(LogMessageType.TECHNICAL,i + ") " + getEntityName(members.get(i)));
        }
    }

    public void logStrengthIncrease(Organism luckyOne) {
        String message = String.format("Strength of the %s has been increased by 3, new strength is %d",
                getEntityName(luckyOne), luckyOne.getStrength());
        log(LogMessageType.ABILITY, message + "\n\n");
        System.out.println(luckyOne.getStrength() + " > " + luckyOne.getStartingStrength());
    }

    public void logOverpopulation(Organism unluckyOne, int neighbours) {
        String message = String.format("%s has died because of overpopulation, it had %d neighbours\n\n",
                getEntityName(unluckyOne), neighbours);
        log(LogMessageType.DEATH,message);
    }

    public void necroLog(Organism rip) {
        String message = String.format("%s is dead :( It was %d rounds old.\n\n",
                getEntityName(rip), rip.getWorld().getRound() - rip.getBirth());
        log(LogMessageType.DEATH, "%s died at (y = %d, x = %d), it was %d rounds old.\n\n",
                getEntityName(rip), rip.getPosition().y, rip.getPosition().x, rip.getWorld().getRound() - rip.getBirth());
    }

    public enum LogMessageType {
        NORMAL(Color.BLACK),
        DEATH(Color.RED),
        SPAWN(Color.GREEN),
        FIGHT(Color.ORANGE),
        WARNING(Color.CYAN),
        ABILITY(Color.BLUE),
        TECHNICAL(Color.MAGENTA);

        private final Color color;

        LogMessageType(Color color) {
            this.color = color;
        }

        public Color getColor() {
            return color;
        }
    }
}
